polkadot/runtime/common/src/impls.rs

```diff
@@ -156,6 +156,27 @@ impl TryConvert<&VersionedLocation, xcm::latest::Location> for VersionedLocation
 	}
 }
 
+/// Adapter for [`Contains`] trait to match [`VersionedLocatableAsset`] type converted to the latest
+/// version of itself where it's location matched by `L` and it's asset id by `A` parameter types.
+pub struct ContainsLatest<L, A>(core::marker::PhantomData<(L, A)>);
+impl<L, A> Contains<VersionedLocatableAsset> for ContainsLatest<L, A>
+where
+	L: Contains<xcm::latest::prelude::Location>,
+	A: Contains<xcm::latest::prelude::Location>,
+{
+	fn contains(asset: &VersionedLocatableAsset) -> bool {
+		use VersionedLocatableAsset::*;
+		let (location, asset_id) = match asset.clone() {
+			V3 { location, asset_id } => match (location.try_into(), asset_id.try_into()) {
+				(Ok(l), Ok(a)) => (l, a),
+				_ => return false,
+			},
+			V4 { location, asset_id } => (location, asset_id),
+		};
+		L::contains(&location) && A::contains(&asset_id.0)
+	}
+}
```

Comment:
I think we should rename `ContainsLatest` to something specific, because it is not that clear if `latest` is related to the used XCM types or `lastest` version of `VersionedLocatableAsset` see also doc `type converted to the latest version of itself`. `VersionedLocatableAsset` has just V3/V4. When we add `xcm:V5` it would be even more confusing.

I would suggest to rename `ContainsLatest` to the `ContainsAssetFrom` and change order of L/A types, e.g.:

```suggestion
pub struct ContainsAssetFrom<A, L>(core::marker::PhantomData<(A, L)>);
impl<A, L> Contains<VersionedLocatableAsset> for ContainsAssetFrom<A, L>
where
	A: Contains<xcm::latest::prelude::Location>,
	L: Contains<xcm::latest::prelude::Location>,
{
	fn contains(asset: &VersionedLocatableAsset) -> bool {
		use VersionedLocatableAsset::*;
		let (location, asset_id) = match asset.clone() {
			V3 { location, asset_id } => match (location.try_into(), asset_id.try_into()) {
				(Ok(l), Ok(a)) => (l, a),
				_ => return false,
			},
			V4 { location, asset_id } => (location, asset_id),
		};
		L::contains(&location) && A::contains(&asset_id.0)
	}
}
```

Reply:
also maybe types:
```
        A: Contains<xcm::latest::prelude::Location>,
	L: Contains<xcm::latest::prelude::Location>,
```
could be better replaced with: `ContainsPair<Location, Location>` or `ContainsPair<AssetId, Location>`